File: android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/events/MapChangeEvent.java
package com.mapbox.rctmgl.events;

import android.view.View;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableMap;

import com.mapbox.rctmgl.events.constants.EventKeys;

public class MapChangeEvent extends AbstractEvent {
    private WritableMap mPayload;

    public MapChangeEvent(View view, String eventType) {
        this(view, eventType, Arguments.createMap());
    }

    public MapChangeEvent(View view, String eventType, WritableMap payload) {
        super(view, eventType);
        mPayload = payload;
    }

    @Override
    public String getKey() {
        return EventKeys.MAP_ONCHANGE;
    }

    @Override
    public WritableMap getPayload() {
        // FMTODO
        WritableMap payloadClone = Arguments.createMap();
        payloadClone.merge(mPayload);
        return payloadClone;
    }

    @Override
    public boolean canCoalesce() {
        // Make sure EventDispatcher never merges EventKeys.MAP_ONCHANGE events.
        // This event name is used to emit events with different
        // com.mapbox.rctmgl.events.constants.EventTypes which are dispatched separately on
        // the JS side
        return false;
    }
}

File: android/src/main/java/com/rnmapbox/rnmbx/events/MapChangeEvent.kt
package com.rnmapbox.rnmbx.events

import android.view.View
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import com.rnmapbox.rnmbx.events.constants.EventKeys

open
class MapChangeEvent constructor(
    view: View,
    eventType: String,
    private val mPayload: WritableMap = Arguments.createMap()
) : AbstractEvent(view, eventType) {
    override val key: String
        get() = EventKeys.MAP_ONCHANGE.value

    override val payload: WritableMap
        get() {
            val payloadClone = Arguments.createMap()
            payloadClone.merge(mPayload)
            return payloadClone
        }

    override fun canCoalesce(): Boolean {
        // Make sure EventDispatcher never merges EventKeys.MAP_ONCHANGE events.
        // This event name is used to emit events with different
        // com.rnmapbox.rnmbx.events.constants.EventTypes which are dispatched separately on
        // the JS side
        return false
    }
}

public
class CameraChangeEvent(view: View, eventType: String, payload: WritableMap) : MapChangeEvent(view, eventType, payload) {
    override fun canCoalesce(): Boolean {
        return true
    }
}